x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ProactiveStorageDeciderService.java

```diff
@@ -0,0 +1,181 @@
+import org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings;
+import org.elasticsearch.cluster.routing.allocation.decider.AllocationDeciders;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingCapacity;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderContext;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderResult;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderService;
+import org.elasticsearch.xpack.core.DataTier;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+public class ProactiveStorageDeciderService implements AutoscalingDeciderService {
+    public static final String NAME = "proactive_storage";
+    public static final Setting<TimeValue> FORECAST_WINDOW = Setting.timeSetting("forecast_window", TimeValue.timeValueMinutes(30));
+
+    private final DiskThresholdSettings diskThresholdSettings;
+    private final AllocationDeciders allocationDeciders;
+
+    public ProactiveStorageDeciderService(Settings settings, ClusterSettings clusterSettings, AllocationDeciders allocationDeciders) {
+        this.diskThresholdSettings = new DiskThresholdSettings(settings, clusterSettings);
+        this.allocationDeciders = allocationDeciders;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public List<DiscoveryNodeRole> roles() {
+        return List.of(DiscoveryNodeRole.DATA_ROLE, DataTier.DATA_HOT_NODE_ROLE);
+    }
+
+    @Override
+    public AutoscalingDeciderResult scale(Settings configuration, AutoscalingDeciderContext context) {
+        AutoscalingCapacity autoscalingCapacity = context.currentCapacity();
+        if (autoscalingCapacity == null || autoscalingCapacity.total().storage() == null) {
+            return new AutoscalingDeciderResult(
+                null,
+                new ReactiveStorageDeciderService.ReactiveReason("current capacity not available", -1, -1)
+            );
+        }
+
+        ReactiveStorageDeciderService.AllocationState allocationState = new ReactiveStorageDeciderService.AllocationState(
+            context,
+            diskThresholdSettings,
+            allocationDeciders
+        );
+        long unassignedBeforeForecast = allocationState.storagePreventsAllocation();
+        assert unassignedBeforeForecast >= 0;
+
+        TimeValue forecastWindow = FORECAST_WINDOW.get(configuration);
+        allocationState = allocationState.forecast(forecastWindow.millis(), System.currentTimeMillis());
```

Comment:
Rather than reassinging a local variable (🙅‍♀️), how about creating a new local variable named `allocationStateAfterForecast` and using that for the rest of the method?

Reply:
ok, done in 232f250fbfe2481d4b8bbf70ddf7484d2ce32bc6